components/browser/menu/src/main/java/mozilla/components/browser/menu/WebExtensionBrowserMenu.kt

```diff
@@ -0,0 +1,95 @@
+import mozilla.components.browser.state.selector.selectedTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.webextension.BrowserAction
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+import java.util.concurrent.ConcurrentHashMap
+
+typealias WebExtensionBrowserAction = BrowserAction
+
+/**
+ * A popup menu for [WebExtensionBrowserAction] menu items.
+ */
+class WebExtensionBrowserMenu internal constructor(
+    adapter: BrowserMenuAdapter,
+    private val store: BrowserStore
+) : BrowserMenu(adapter) {
+    private var scope: CoroutineScope? = null
+
+    @kotlinx.coroutines.ExperimentalCoroutinesApi
+    override fun show(
+        anchor: View,
+        orientation: Orientation,
+        endOfMenuAlwaysVisible: Boolean,
+        onDismiss: () -> Unit
+    ): PopupWindow {
+        scope = store.flowScoped { flow ->
+            flow.ifChanged { it.selectedTab }
+                .collect { state ->
+                    getOrUpdateWebExtensionMenuItems(state, state.selectedTab)
+                    invalidate()
+                }
+        }
+
+        return super.show(
+            anchor,
+            orientation,
+            endOfMenuAlwaysVisible,
+            onDismiss
+        ).apply {
+            setOnDismissListener {
+                adapter.menu = null
+                currentPopup = null
+                scope?.cancel()
+                webExtensionBrowserActions.clear()
+                onDismiss()
+            }
+        }
+    }
+
+    companion object {
+        internal val webExtensionBrowserActions = ConcurrentHashMap<String, WebExtensionBrowserMenuItem>()
+
+        internal fun getOrUpdateWebExtensionMenuItems(
+            state: BrowserState,
+            tab: SessionState? = null
+        ): List<BrowserMenuItem> {
+            val menuItems = ArrayList<BrowserMenuItem>()
+            val extensions = state.extensions.values.toList()
+            extensions.filter { it.enabled }.forEach { extension ->
+                extension.browserAction?.let { browserAction ->
+                    // Add the global browser action if it doesn't exist
+                    val browserMenuItem = webExtensionBrowserActions.getOrPut(extension.id) {
+                        val browserMenuItem = WebExtensionBrowserMenuItem(
+                            browserAction = browserAction,
+                            listener = browserAction.onClick
+                        )
+                        browserMenuItem
+                    }
+
+                    tab?.extensionState?.get(extension.id)?.browserAction?.let {
+                        browserMenuItem.browserAction = browserMenuItem.browserAction.copyWithOverride(it)
```

Comment:
This needs to be `browserMenuItem.browserAction = browserAction.copyWithOverride(it)` -> We want to start from the global one and applying the overrides of the tab one.